EIPS/eip-6900.md

```diff
@@ -88,190 +79,334 @@ Each step is modular, supporting different implementations for each execution fu
+    /// @param functionId An identifier that routes the call to different internal implementations, should there be more than one.
+    /// @param userOp The user operation.
+    /// @param userOpHash The user operation hash.
+    /// @return Packed validation data for validAfter (6 bytes), validUntil (6 bytes), and authorizer (20 bytes).
+    function userOpValidator(uint8 functionId, UserOperation memory userOp, bytes32 userOpHash) external returns (uint256);
+
+    /// @notice Run the pre runtime validation hook specified by the `functionId`.
+    /// @dev To indicate the entire call should revert, the function MUST revert.
+    /// @param functionId An identifier that routes the call to different internal implementations, should there be more than one.
+    /// @param sender The caller address.
+    /// @param value The call value.
+    /// @param data The calldata sent.
+    function preRuntimeValidationHook(uint8 functionId, address sender, uint256 value, bytes calldata data) external;
+
+    /// @notice Run the runtime validator specified by the `functionId`.
+    /// @dev To indicate the entire call should revert, the function MUST revert.
+    /// @param functionId An identifier that routes the call to different internal implementations, should there be more than one.
+    /// @param sender The caller address.
+    /// @param value The call value.
+    /// @param data The calldata sent.
+    function runtimeValidator(uint8 functionId, address sender, uint256 value, bytes calldata data) external;
+
+    /// @notice Run the pre execution hook specified by the `functionId`.
+    /// @dev To indicate the entire call should revert, the function MUST revert.
+    /// @param functionId An identifier that routes the call to different internal implementations, should there be more than one.
+    /// @param sender The caller address.
+    /// @param value The call value.
+    /// @param data The calldata sent.
+    /// @return Context to pass to a post execution hook, if present. An empty bytes array MAY be returned.
+    function preExecutionHook(uint8 functionId, address sender, uint256 value, bytes calldata data) external returns (bytes memory);
+
+    /// @notice Run the post execution hook specified by the `functionId`.
+    /// @dev To indicate the entire call should revert, the function MUST revert.
+    /// @param functionId An identifier that routes the call to different internal implementations, should there be more than one.
+    /// @param preExecHookData The context returned by its associated pre execution hook.
+    function postExecutionHook(uint8 functionId, bytes calldata preExecHookData) external;
+
+    /// @notice Describe the contents and intended configuration of the plugin.
+    /// @dev This function MUST NOT make any state changes.
+    /// @return A manifest describing the contents and intended configuration of the plugin.
+    function pluginManifest() external pure returns (PluginManifest memory);
```

Comment:
`pure` indicates that it does not read any state at all, would that mean that the manifest data is kept in the immutable storage (I think this is not directly possible for structs right now)? Also this does not align with  `This function MUST NOT make any state changes.`, which is the description for a `view` function.

Reply:
Thanks for the thorough review!

Manifest data should be declared within the `pluginManifest()` function itself. With plugin dependencies provided at installation time, the plugin does not need to read from storage.

Agree that `/// @dev This function MUST NOT make any state changes.` is redundant / not apt here given `pure`.

The manifest hash check during installation is meant to prevent installation of malicious plugins that may swap its manifest to a different one in between user confirmation and actual execution.

Reply:
Thanks for the explanation, this makes sense :)

Since there is a manifest hash check it is not super critical that this is actually a pure function, right? Since a change to the manifest would be detected.

Reply:
That's correct. I do like `pure` in the sense that it conveys to the developer that what should be returned should be fixed and not change over time (and possibly read from mutable state). However, I do see the possibility that `pure` might be difficult to work with in certain scenarios.

Do you see practical benefits to this being relaxed to `view`?

Reply:
> Do you see practical benefits to this being relaxed to view?

Good question: My comment was primarily because I was confused by the slight mismatch of natspec and modifier. 

Personally I would probably relax it as it is not a real security benefit (due to the hash check). This will give the developer the flexibility to store the data however they want. That being said they can already do this by "just not following the interface 100%", as there is no opcode to enforce `pure` (like staticall for `view`). Also I cannot provide a concrete use case were storing it in memory would be required, therefore not strong preference ;)

Reply:
Got it thanks for your thoughts! Will continue to iterate on this.